zpages/window_stat.go

```diff
@@ -0,0 +1,124 @@
+package zpages
+
+import (
+	"math"
+	"time"
+
+	"go.opencensus.io/stats/view"
+)
+
+type interval struct {
+	distribution *view.DistributionData
+	index        int64
+	updateTime   time.Time
+}
+
+type windowStat struct {
+	intervals      []interval
+	lastUpdate     int
+	granularitySec int64
+	startUnix      int64
+	startTime      time.Time
+	window         time.Duration
+}
+
+var clock = func() time.Time {
+	return time.Now()
+}
+
+func (ws *windowStat) init(window time.Duration) {
+	historySec := int64(window.Seconds())
+	granularitySec := int64(historySec / 60)
+	ws.intervals = make([]interval, int(historySec/granularitySec+1))
+	for i := range ws.intervals {
+		ws.intervals[i].index = -1
+	}
+	ws.granularitySec = granularitySec
+	ws.lastUpdate = -1
+	ws.window = window
+}
+
+func (ws *windowStat) update(t time.Time, dist *view.DistributionData) {
+	if ws.lastUpdate == -1 {
+		ws.startTime = t
+		ws.startUnix = t.Unix() - (t.Unix() % ws.granularitySec)
+	}
+	sliceIndex, intv := ws.intervalFor(t)
+	if intv != nil {
+		intv.distribution = dist
+		intv.updateTime = t
+		ws.lastUpdate = sliceIndex
+	}
+}
+
+func (ws *windowStat) intervalFor(t time.Time) (int, *interval) {
+	deltaSec := t.Unix() - ws.startUnix
+	if deltaSec < 0 {
+		return -1, nil
+	}
+	ind := deltaSec / ws.granularitySec
+	n := int64(len(ws.intervals))
+	sliceIndex := int(ind % n)
+	intv := &ws.intervals[sliceIndex]
+	if intv.index == ind {
+		return sliceIndex, intv
+	}
+	if intv.index < ind {
+		*intv = interval{index: ind}
+		return sliceIndex, intv
+	}
+	// we are attempting to get an interval for something that has already
+	// fallen out of the circular buffer
+	return -1, nil
+}
+
+func (ws *windowStat) read() (actualStart, actualEnd time.Time, diff *view.DistributionData) {
```

Comment:
Is this all newly written code or is it just reverted code? I don't have too much context about it, hence why am asking.

Reply:
It is new